File: apps/websocket-skyblock-csharp/src/hypixel/routes/BazaarRoute/BazaarRoute.cs
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using Newtonsoft.Json;

public class BazaarRoute
{
    public bool success = default!;
    public double lastUpdated = default!;
    public Dictionary<string, BazaarRouteProduct> products = new Dictionary<string, BazaarRouteProduct>();

    public static async Task<Dictionary<string, BazaarRouteProduct>?> GetRoute()
    {
        try
        {
            string json = "";
            int maxRetry = 3;
            int currentRetry = 0;

            async Task setJson() => json = await Program.Client.GetStringAsync($"{Settings.HYPIXEL_API_BASE_URL}/skyblock/bazaar");

            try
            {
                await setJson();
            }

            catch (Exception)
            {
                currentRetry++;
                Thread.Sleep(100);

                Utility.Log(Enums.LogLevel.WARN, $"Fetch failed! ({currentRetry})");
                if (currentRetry >= maxRetry)
                    throw;

                await setJson();
            }

            JsonNode? result = JsonArray.Parse(json);

            if (result == null)
            {
                Utility.Log(Enums.LogLevel.ERROR, "Unexpected null when the result should not be null.");
                return null;
            }

            if (result["success"] == null)
            {
                Utility.Log(Enums.LogLevel.ERROR, "Unexpected null when the result should always contains a 'success' field.");
                return null;
            }

            if (result["success"]!.Equals(false))
            {
                Utility.Log(Enums.LogLevel.WARN, "Success = false, returning...");
                return null;
            }

            BazaarRoute? bazaarRoute = JsonConvert.DeserializeObject<BazaarRoute>(json);

            if (bazaarRoute == null)
            {
                Utility.Log(Enums.LogLevel.ERROR, "Unexpected null while converting response to C# class.");
                return null;
            }

            return bazaarRoute.products;
        }

        catch (Exception e)
        {
            Utility.Log(Enums.LogLevel.ERROR, e.ToString());
            return null;
        }
    }
}

File: apps/discord-bot/app/src/main/java/discordbot/hypixel/routes/BazaarRoute/BazaarRoute.java
package discordbot.hypixel.routes.BazaarRoute;

import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse.BodyHandlers;
import java.text.MessageFormat;
import java.util.HashMap;

import discordbot.BotSettings;
import discordbot.Main;
import discordbot.common.Enums.LogLevel;
import discordbot.common.Enums.Settings;

public class BazaarRoute {
    public boolean success;
    public double lastUpdated;
    public HashMap<String, BazaarRouteProduct> products = new HashMap<String, BazaarRouteProduct>();

    public static HashMap<String, BazaarRouteProduct> GetRoute() {
        String json = new String();
        int maxRetry = 3;
        int currentRetry = 0;
        HashMap<String, BazaarRouteProduct> products = new HashMap<String, BazaarRouteProduct>();

        try {
            do {
                try {
                    URI uri = new URI(MessageFormat.format("{0}/skyblock/bazaar", BotSettings.Get(Settings.HYPIXEL_API_BASE_URL)));
                    HttpRequest request = HttpRequest.newBuilder(uri).header("API-Key", BotSettings.Get(Settings.HYPIXEL_BOT_KEY)).build();
                    json = Main.Client.send(request, BodyHandlers.ofString()).body();
                    BazaarRoute product = Main.Gson.fromJson(json, BazaarRoute.class);
                    products = product.products;
                    currentRetry = maxRetry + 1;
                } catch (Exception e) {
                    Thread.sleep(100);
                    currentRetry += 1;
                    Main.Utility.Log(LogLevel.WARN, MessageFormat.format("Fetch failed! ({0})", e));

                    if (currentRetry >= maxRetry)
                        throw new Exception(MessageFormat.format("Reached max retry limit! Throwing... {0}", e));
                }
            } while (currentRetry <= maxRetry);
        } catch (Exception err) {
            Main.Utility.Log(LogLevel.ERROR, MessageFormat.format("Unhandled Exception: {0}", err));
        }

        return products;
    }
}